File: py/structure/tree/binary_tree/BinaryTree.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class TreeNode:
    def __init__(self, value: int):
        self.value, self.left, self.right = value, None, None

class BinaryTree:
    def isPerfect(self, root: TreeNode) -> bool:
        if not root:
            return True

        queue = [root]
        count, flag = 0, False
        while queue:
            current = queue.pop(0)
            if not current:
                flag = True
                continue

            if flag:
                return False

            count += 1
            queue.append(current.left)
            queue.append(current.right)
        return not (count & (count + 1))

    def isCompleted(self, root: TreeNode) -> bool:
        if not root:
            return True

        queue = [root]
        flag = False
        while queue:
            current = queue.pop(0)
            if (not current.left and current.right) or (flag and (current.left or current.right)):
                return False

            if current.left:
                queue.append(current.left)
            if current.right:
                queue.append(current.right)
            else:
                flag = True
        return True

    def isFull(self, root: TreeNode) -> bool:
        if not root:
            return True

        queue = [root]
        while queue:
            current = queue.pop(0)
            if (current.left and not current.right) or (not current.left and current.right):
                return False

            if current.left:
                queue.append(current.left)
            if current.right:
                queue.append(current.right)
        return True

    def nodeNum(self, root: TreeNode) -> int:
        if not root:
            return 0

        queue = [root]
        count = 0
        while queue:
            count += 1

            current = queue.pop(0)
            if current.left:
                queue.append(current.left)
            if current.right:
                queue.append(current.right)
        return count

    def leafNodeNum(self, root: TreeNode) -> int:
        if not root:
            return 0

        queue = [root]
        count = 0
        while queue:
            current = queue.pop(0)
            if not current.left and not current.right:
                count += 1
                continue

            if current.left:
                queue.append(current.left)
            if current.right:
                queue.append(current.right)
        return count

    def maxDepth(self, root: TreeNode) -> int:
        if not root:
            return 0

        queue = [root]
        count = 0
        while queue:
            count += 1

            length = len(queue)
            for _ in range(length):
                current = queue.pop(0)
                if current.left:
                    queue.append(current.left)
                if current.right:
                    queue.append(current.right)
        return count

    def mirror(self, root: TreeNode) -> TreeNode:
        if not root or (not root.left and not root.right):
            return root

        root.left, root.right = root.right, root.left
        root.left, root.right = self.mirror(root.left), self.mirror(root.right)
        return root

File: cpp/structure/tree/binary_tree/BinaryTree.cpp
#include <iostream>
#include <queue>
using namespace std;

struct TreeNode {
    int value;
    TreeNode *left;
    TreeNode *right;

    TreeNode(int value) : value(value), left(nullptr), right(nullptr) {}
};

class BinaryTree {
public:
    bool isPerfect(TreeNode *root) {
        if (!root) {
            return true;
        }

        queue<TreeNode *> queue;
        queue.push(root);

        int count = 0;
        bool flag = false;
        while (!queue.empty()) {
            TreeNode *current = queue.front();
            queue.pop();
            if (!current) {
                flag = true;
                continue;
            }

            if (flag) {
                return false;
            }

            ++count;
            queue.push(current->left);
            queue.push(current->right);
        }
        return !(count & (count + 1));
    }

    bool isCompleted(TreeNode *root) {
        if (!root) {
            return true;
        }

        queue<TreeNode *> queue;
        queue.push(root);

        bool flag = false;
        while (!queue.empty()) {
            TreeNode *current = queue.front();
            queue.pop();
            if ((!current->left && current->right) || (flag && (current->left || current->right))) {
                return false;
            }

            if (current->left) {
                queue.push(current->left);
            }
            if (current->right) {
                queue.push(current->right);
            } else {
                flag = true;
            }
        }
        return true;
    }

    bool isFull(TreeNode *root) {
        if (!root) {
            return true;
        }

        queue<TreeNode *> queue;
        queue.push(root);
        while (!queue.empty()) {
            TreeNode *current = queue.front();
            queue.pop();
            if ((current->left && !current->right) || (!current->left && current->right)) {
                return false;
            }

            if (current->left) {
                queue.push(current->left);
            }
            if (current->right) {
                queue.push(current->right);
            }
        }
        return true;
    }

    int nodeNum(TreeNode *root) {
        if (!root) {
            return 0;
        }

        queue<TreeNode *> queue;
        queue.push(root);

        int count = 0;
        while (!queue.empty()) {
            ++count;

            TreeNode *current = queue.front();
            queue.pop();
            if (current->left) {
                queue.push(current->left);
            }
            if (current->right) {
                queue.push(current->right);
            }
        }
        return count;
    }

    int leafNodeNum(TreeNode *root) {
        if (!root) {
            return 0;
        }

        queue<TreeNode *> queue;
        queue.push(root);

        int count = 0;
        while (!queue.empty()) {
            TreeNode *current = queue.front();
            queue.pop();
            if (!current->left && !current->right) {
                ++count;
                continue;
            }

            if (current->left) {
                queue.push(current->left);
            }
            if (current->right) {
                queue.push(current->right);
            }
        }
        return count;
    }

    int maxDepth(TreeNode *root) {
        if (!root) {
            return 0;
        }

        queue<TreeNode *> queue;
        queue.push(root);

        int count = 0;
        while (!queue.empty()) {
            ++count;

            int length = queue.size();
            for (int i = 0; i < length; ++i) {
                TreeNode *current = queue.front();
                queue.pop();
                if (current->left) {
                    queue.push(current->left);
                }
                if (current->right) {
                    queue.push(current->right);
                }
            }
        }
        return count;
    }

    TreeNode *mirror(TreeNode *root) {
        if (!root || (!root->left && !root->right)) {
            return root;
        }

        swap(root->left, root->right);

        root->left = mirror(root->left);
        root->right = mirror(root->right);
        return root;
    }
};
